changefeedccl: use parallel event preprocessor

jayshrivastava opened the pull request.

Commit:
span: make span.Frontier thread safe

Make `span.Frontier` thread safe by default.

Release justification: low impact data structure change.
Release note: None

jayshrivastava committed:
changefeedccl: separate preprocessing and emitting of events

This change refactors the event consumer to encode and
emit events using separate functions. This effectively
separates the parallelizable and nonparallelizable parts
of consuming events.

Release justification: Small change which preserves all
existing functionality.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87225)
<!-- Reviewable:end -->

jayshrivastava committed:
changefeedccl: implement parallel preprocessor

This change implements a parallel preprocessor which
fetches events, encodes them, and buffers them until
they are consumed.

The parallel preprocessor is implemented with 3 stages.

The `producerLoop` fetches events from the blocking buffer,
shards them by key, and passes them to workers. There can
only be one `producerLoop` running to preserve per-key
ordering guarantees.

A `workerLoop` takes each event and 'preprocesses' it,
which mainly means encoding. It uses the underlying
`eventConsumer` which has functions for preprocessing.
The number of workers is configurable.

Finally, `GetNextEvent` fetches an event and a `consume`
callback function. The callback function implicitly buffers
an encoded event and will emit the encoded event
when it is called.

This change also adds tests which verify per-key ordering
guarantees and graceful termination due to context cancellation
or calling `Done`.

Release justification: This change adds unused code.

Release note: None

jayshrivastava committed:
changefeedccl: add cluster setting to control event preprocessing

Previously, an aggregator would synchronously fetch an event, encode it,
and emit it. This was a bottleneck to throughput.

This change introduces the parallelPreprocessor struct which
has workers fetch and encode events in the background. These workers buffer
the events until they are ready to be emitted, which still happens
synchronously.

This change is hidden behind a cluster setting,
changefeed.num_event_preprocessors, which is set to 0 by default, disabling
preprocessing. After benchmarking this change, we may set it to a nonzero
number in the future. This change also adds the setting
changefeed.event_preprocessor_worker_queue_size, which controls the number of
events each worker buffers.

Release justification: This change adds a feature which may have
significant benefits to changefeed performance. A cluster setting
has been added to disable these changes by default, which makes this
change low risk.

Release note (enterprise change): New cluster settings
changefeed.num_event_preprocessors and
changefeed.event_preprocessor_worker_queue_size can be used to
add parallel encoding to cdc events.

num_event_preprocessors controls the number of concurrent workers per
node which can work to pre-encode events before they are emitted.

event_preprocessor_worker_queue_size controls the max number of events
each worker can buffer.

jayshrivastava committed:
do not merge: force 32 event preprocessors for testing in ci

Release justification:

jayshrivastava commented:
Decided not to move forward with this due it having complexity and low parallelism. Looking to continue with the async consumer implementation in https://github.com/jayshrivastava/cockroach/tree/nprocs-3
